File: marketStateAnalyzer.js
// Faithful port of marketStateAnalyzer.ts. Every formula below was copied
// directly from the source file, not reconstructed from memory - includes
// full MACD/RSI divergence detection.

const { ema, rsi, macd, atrRatio, macdHistogramTurn, avgVolume } = require("./indicators");

const OVERSOLD_EXTREME_THRESHOLD = 20;
const OVERSOLD_MILD_THRESHOLD = 30;
const OVERBOUGHT_EXTREME_THRESHOLD = 80;
const OVERBOUGHT_MILD_THRESHOLD = 70;

function buildTimeframeIndicators(candles) {
  const closes = candles.map((c) => c.close);
  const currentPrice = closes[closes.length - 1];
  const ema20Arr = ema(closes, 20);
  const ema50Arr = ema(closes, 50);
  const ema20 = ema20Arr[ema20Arr.length - 1] || 0;
  const ema50 = ema50Arr[ema50Arr.length - 1] || 0;
  const m = macd(closes) || { macd: 0, signal: 0, histogram: 0 };
  // One period back, for MACD-histogram-reversal detection (fixes a gap in
  // the original where this was never computed at all).
  const prevM = closes.length > 1 ? macd(closes.slice(0, -1)) : null;
  const deviationFromEMA20 = ema20 !== 0 ? ((currentPrice - ema20) / ema20) * 100 : 0;
  return {
    currentPrice,
    ema20,
    ema50,
    macd: m.macd,
    macdSignal: m.signal,
    macdHistogram: m.histogram,
    prevMacdHistogram: prevM ? prevM.histogram : undefined,
    macdTurn: macdHistogramTurn(closes),
    rsi7: rsi(closes, 7) ?? 50,
    rsi14: rsi(closes, 14) ?? 50,
    atrRatio: atrRatio(candles, 14),
    deviationFromEMA20,
    volume: candles[candles.length - 1].volume,
    avgVolume: avgVolume(candles, 20),
    candles,
  };
}

function determineTrendStrength(tf) {
  if (tf.ema20 > tf.ema50 && tf.macd > 0) return "trending_up";
  if (tf.ema20 < tf.ema50 && tf.macd < 0) return "trending_down";
  return "ranging";
}

function determineMomentumState(tf) {
  if (tf.rsi7 < OVERSOLD_EXTREME_THRESHOLD) return "oversold_extreme";
  if (tf.rsi7 < OVERSOLD_MILD_THRESHOLD) return "oversold_mild";
  if (tf.rsi7 > OVERBOUGHT_EXTREME_THRESHOLD) return "overbought_extreme";
  if (tf.rsi7 > OVERBOUGHT_MILD_THRESHOLD) return "overbought_mild";
  return "neutral";
}

function determineVolatilityState(tf) {
  if (tf.atrRatio > 1.5) return "high_vol";
  if (tf.atrRatio < 0.7) return "low_vol";
  return "normal_vol";
}

function determineMarketState(trendStrength, momentumState, tf15m) {
  let state = "no_clear_signal";
  let confidence = 0.3;

  if (trendStrength === "trending_up" && momentumState === "oversold_extreme") {
    state = "uptrend_oversold"; confidence = 0.9;
  } else if (trendStrength === "trending_down" && momentumState === "overbought_extreme") {
    state = "downtrend_overbought"; confidence = 0.9;
  } else if (trendStrength === "trending_down" && momentumState === "oversold_extreme") {
    state = "downtrend_oversold"; confidence = 0.6;
  } else if (trendStrength === "trending_up" && momentumState === "overbought_extreme") {
    state = "uptrend_overbought"; confidence = 0.6;
  } else if (trendStrength === "trending_up" && (momentumState === "oversold_mild" || momentumState === "neutral")) {
    state = "uptrend_continuation"; confidence = 0.7;
  } else if (trendStrength === "trending_down" && (momentumState === "overbought_mild" || momentumState === "neutral")) {
    state = "downtrend_continuation"; confidence = 0.7;
  } else if (trendStrength === "trending_down" && momentumState === "oversold_mild") {
    state = "downtrend_oversold"; confidence = 0.5;
  } else if (trendStrength === "trending_up" && momentumState === "overbought_mild") {
    state = "uptrend_overbought"; confidence = 0.5;
  } else if (trendStrength === "ranging" && momentumState === "oversold_extreme") {
    state = "ranging_oversold"; confidence = 0.8;
  } else if (trendStrength === "ranging" && momentumState === "overbought_extreme") {
    state = "ranging_overbought"; confidence = 0.8;
  } else if (trendStrength === "ranging" && momentumState === "neutral") {
    state = "ranging_neutral"; confidence = 0.5;
  }

  if (tf15m.macdTurn === 1 && (state === "uptrend_oversold" || state === "ranging_oversold")) {
    confidence = Math.min(confidence + 0.1, 1.0);
  }
  if (tf15m.macdTurn === -1 && (state === "downtrend_overbought" || state === "ranging_overbought")) {
    confidence = Math.min(confidence + 0.1, 1.0);
  }

  return { state, confidence };
}

function calculateTrendConsistency(ema20A, ema50A, ema20B, ema50B, macdA, macdB) {
  let score = 0;
  const trendA = ema20A > ema50A ? 1 : -1;
  const momentumA = macdA > 0 ? 1 : -1;
  const trendB = ema20B > ema50B ? 1 : -1;
  const momentumB = macdB > 0 ? 1 : -1;
  if (trendA === trendB) score += 0.4;
  if (momentumA === momentumB) score += 0.3;
  if (trendA === momentumA) score += 0.15;
  if (trendB === momentumB) score += 0.15;
  return Math.max(0, Math.min(1, score));
}

function calculateTripleTimeframeConsistency(tfPrimary, tfConfirm, tfFilter) {
  const pc = calculateTrendConsistency(tfPrimary.ema20, tfPrimary.ema50, tfConfirm.ema20, tfConfirm.ema50, tfPrimary.macd, tfConfirm.macd);
  const cf = calculateTrendConsistency(tfConfirm.ema20, tfConfirm.ema50, tfFilter.ema20, tfFilter.ema50, tfConfirm.macd, tfFilter.macd);
  return pc * 0.6 + cf * 0.4;
}

// Trend score (-100..100): EMA gap (40%) + MACD/price normalized (30%) +
// price deviation from EMA20 (20%) + RSI trend (10%). Matches source exactly.
function calculateTrendScore(tf) {
  let score = 0;
  const emaGap = (tf.ema20 - tf.ema50) / tf.ema50;
  score += Math.max(-40, Math.min(40, emaGap * 1000));
  const macdNormalized = tf.macd / tf.currentPrice;
  score += Math.max(-30, Math.min(30, macdNormalized * 10000));
  score += Math.max(-20, Math.min(20, tf.deviationFromEMA20 * 2));
  const rsiTrend = (tf.rsi7 - 50) / 5;
  score += Math.max(-10, Math.min(10, rsiTrend));
  return Math.round(score);
}

// Matches source exactly: 20%-relative-drop weakening, +/-20 crossing = reversing.
function detectTrendWeakening(currentScore, scoreHistory) {
  const previousScore = scoreHistory.length > 0 ? scoreHistory[scoreHistory.length - 1] : currentScore;
  const change = currentScore - previousScore;
  const changePercent = previousScore !== 0 ? (change / Math.abs(previousScore)) * 100 : 0;
  const isWeakening = Math.abs(currentScore) < Math.abs(previousScore) * 0.8;
  const isReversing = (previousScore > 20 && currentScore < -20) || (previousScore < -20 && currentScore > 20);
  const weakeningSeverity = isWeakening ? Math.round((1 - Math.abs(currentScore) / Math.abs(previousScore)) * 100) : 0;
  return { currentScore, previousScore, change, changePercent, isWeakening, isReversing, weakeningSeverity };
}

// ---- MACD / RSI divergence detection (matches source exactly) ----

function detectMACDDivergence(candles, macdValues) {
  if (!candles || candles.length < 20 || !macdValues || macdValues.length < 20) {
    return { type: "none", strength: 0, description: "insufficient data" };
  }
  const prices = candles.map((c) => c.close);
  const halfLen = Math.floor(prices.length / 2);
  const firstHalfPrices = prices.slice(0, halfLen);
  const secondHalfPrices = prices.slice(halfLen);
  const firstHalfMACD = macdValues.slice(0, halfLen);
  const secondHalfMACD = macdValues.slice(halfLen);

  const priceHigh1 = Math.max(...firstHalfPrices);
  const priceHigh2 = Math.max(...secondHalfPrices);
  const priceLow1 = Math.min(...firstHalfPrices);
  const priceLow2 = Math.min(...secondHalfPrices);
  const macdHigh1 = Math.max(...firstHalfMACD);
  const macdHigh2 = Math.max(...secondHalfMACD);
  const macdLow1 = Math.min(...firstHalfMACD);
  const macdLow2 = Math.min(...secondHalfMACD);

  const isPriceHigherHigh = priceHigh2 > priceHigh1 * 1.001;
  const isMACDLowerHigh = macdHigh2 < macdHigh1 * 0.95;
  if (isPriceHigherHigh && isMACDLowerHigh) {
    const priceIncrease = ((priceHigh2 - priceHigh1) / priceHigh1) * 100;
    const macdDecrease = ((macdHigh1 - macdHigh2) / Math.abs(macdHigh1)) * 100;
    const strength = Math.min(100, Math.round((priceIncrease + macdDecrease) * 10));
    return { type: "bearish", strength: Math.max(60, strength), description: `bearish MACD divergence: price new high (${priceHigh2.toFixed(2)}) but MACD didn't confirm (${macdHigh2.toFixed(4)})` };
  }

  const isPriceLowerLow = priceLow2 < priceLow1 * 0.999;
  const isMACDHigherLow = macdLow2 > macdLow1 * 1.05;
  if (isPriceLowerLow && isMACDHigherLow) {
    const priceDecrease = ((priceLow1 - priceLow2) / priceLow1) * 100;
    const macdIncrease = ((macdLow2 - macdLow1) / Math.abs(macdLow1)) * 100;
    const strength = Math.min(100, Math.round((priceDecrease + macdIncrease) * 10));
    return { type: "bullish", strength: Math.max(60, strength), description: `bullish MACD divergence: price new low (${priceLow2.toFixed(2)}) but MACD didn't confirm (${macdLow2.toFixed(4)})` };
  }

  return { type: "none", strength: 0, description: "no clear divergence" };
}

function detectRSIDivergence(candles, rsiValues) {
  if (!candles || candles.length < 20 || !rsiValues || rsiValues.length < 20) {
    return { type: "none", strength: 0, description: "insufficient data" };
  }
  const prices = candles.map((c) => c.close);
  const halfLen = Math.floor(prices.length / 2);
  const firstHalfPrices = prices.slice(0, halfLen);
  const secondHalfPrices = prices.slice(halfLen);
  const firstHalfRSI = rsiValues.slice(0, halfLen);
  const secondHalfRSI = rsiValues.slice(halfLen);

  const priceHigh1 = Math.max(...firstHalfPrices);
  const priceHigh2 = Math.max(...secondHalfPrices);
  const priceLow1 = Math.min(...firstHalfPrices);
  const priceLow2 = Math.min(...secondHalfPrices);
  const rsiHigh1 = Math.max(...firstHalfRSI);
  const rsiHigh2 = Math.max(...secondHalfRSI);
  const rsiLow1 = Math.min(...firstHalfRSI);
  const rsiLow2 = Math.min(...secondHalfRSI);

  const isPriceHigherHigh = priceHigh2 > priceHigh1 * 1.001;
  const isRSILowerHigh = rsiHigh2 < rsiHigh1 - 3;
  if (isPriceHigherHigh && isRSILowerHigh) {
    const priceIncrease = ((priceHigh2 - priceHigh1) / priceHigh1) * 100;
    const rsiDecrease = rsiHigh1 - rsiHigh2;
    const strength = Math.min(100, Math.round((priceIncrease * 5 + rsiDecrease) * 2));
    return { type: "bearish", strength: Math.max(60, strength), description: `bearish RSI divergence: price new high (${priceHigh2.toFixed(2)}) but RSI didn't confirm (${rsiHigh2.toFixed(1)})` };
  }

  const isPriceLowerLow = priceLow2 < priceLow1 * 0.999;
  const isRSIHigherLow = rsiLow2 > rsiLow1 + 3;
  if (isPriceLowerLow && isRSIHigherLow) {
    const priceDecrease = ((priceLow1 - priceLow2) / priceLow1) * 100;
    const rsiIncrease = rsiLow2 - rsiLow1;
    const strength = Math.min(100, Math.round((priceDecrease * 5 + rsiIncrease) * 2));
    return { type: "bullish", strength: Math.max(60, strength), description: `bullish RSI divergence: price new low (${priceLow2.toFixed(2)}) but RSI didn't confirm (${rsiLow2.toFixed(1)})` };
  }

  return { type: "none", strength: 0, description: "no clear divergence" };
}

// Reconstructs an approximate historical MACD/RSI series from candle closes,
// matching the source's own approximation approach (it also estimates
// historical MACD/RSI from price ratios rather than true historical recompute).
function estimateMacdSeries(candles, currentMacd, currentPrice) {
  const closes = candles.map((c) => c.close).slice(-30);
  return closes.map((close) => currentMacd * (close / currentPrice));
}
function estimateRsiSeries(candles, currentRsi7) {
  const closes = candles.map((c) => c.close).slice(-30);
  const out = [];
  for (let i = 0; i < closes.length; i++) {
    const idx = closes.length - 1 - i;
    if (idx >= 0 && idx < closes.length - 1) {
      const priceChange = ((closes[idx + 1] - closes[idx]) / closes[idx]) * 100;
      const rsiAdjust = priceChange * 0.5;
      out.unshift(Math.max(0, Math.min(100, currentRsi7 - rsiAdjust)));
    } else {
      out.unshift(currentRsi7);
    }
  }
  return out;
}

// ---- Persisted trend-score history (replaces the original's in-memory Map) ----

function updateHistory(historyStore, symbol, scores) {
  const now = Date.now();
  const CACHE_EXPIRE_MS = 3600000;
  const HISTORY_SIZE = 5;
  let entry = historyStore[symbol];
  if (!entry || now - entry.lastUpdate > CACHE_EXPIRE_MS) {
    entry = { primary: [scores.primary], confirm: [scores.confirm], filter: [scores.filter], lastUpdate: now };
  } else {
    entry.primary.push(scores.primary);
    entry.confirm.push(scores.confirm);
    entry.filter.push(scores.filter);
    if (entry.primary.length > HISTORY_SIZE) {
      entry.primary.shift(); entry.confirm.shift(); entry.filter.shift();
    }
    entry.lastUpdate = now;
  }
  historyStore[symbol] = entry;
  return historyStore;
}

function getHistory(historyStore, symbol) {
  const entry = historyStore[symbol];
  if (!entry || Date.now() - entry.lastUpdate > 3600000) return { primary: [], confirm: [], filter: [] };
  return { primary: entry.primary.slice(), confirm: entry.confirm.slice(), filter: entry.filter.slice() };
}

// Full reversal score (0-100), matching source weighting:
// primary 40% / confirm 25% / filter 15% / MACD divergence 10% / RSI divergence 10%.
function calculateReversalScore(tfPrimary, tfConfirm, tfFilter, positionDirection, history) {
  let score = 0;
  const details = [];
  const reversedFrames = [];

  const scorePrimary = calculateTrendScore(tfPrimary);
  const scoreConfirm = calculateTrendScore(tfConfirm);
  const scoreFilter = calculateTrendScore(tfFilter);
  const targetSign = positionDirection === "long" ? -1 : 1;
  const targetDivergence = positionDirection === "long" ? "bearish" : "bullish";

  const primaryChange = detectTrendWeakening(scorePrimary, history.primary);
  if (Math.sign(scorePrimary) === targetSign && Math.abs(scorePrimary) > 30) {
    score += 40; details.push(`primary timeframe strongly reversed (score=${scorePrimary})`); reversedFrames.push("primary");
  } else if (primaryChange.isWeakening && primaryChange.weakeningSeverity > 40) {
    score += 20; details.push(`primary timeframe weakening significantly (${primaryChange.weakeningSeverity}%)`);
  } else if (Math.abs(scorePrimary) < 20) {
    score += 12; details.push(`primary timeframe entering range (score=${scorePrimary})`);
  }

  const confirmChange = detectTrendWeakening(scoreConfirm, history.confirm);
  if (Math.sign(scoreConfirm) === targetSign && Math.abs(scoreConfirm) > 30) {
    score += 25; details.push(`confirm timeframe strongly reversed (score=${scoreConfirm})`); reversedFrames.push("confirm");
  } else if (confirmChange.isWeakening && confirmChange.weakeningSeverity > 40) {
    score += 12; details.push(`confirm timeframe weakening significantly (${confirmChange.weakeningSeverity}%)`);
  }

  const filterChange = detectTrendWeakening(scoreFilter, history.filter);
  if (Math.sign(scoreFilter) === targetSign && Math.abs(scoreFilter) > 30) {
    score += 15; details.push(`filter timeframe reversed (score=${scoreFilter})`); reversedFrames.push("filter");
  }

  // MACD divergence (10%) - primary timeframe
  if (tfPrimary.candles && tfPrimary.candles.length >= 20) {
    const macdSeries = estimateMacdSeries(tfPrimary.candles, tfPrimary.macd, tfPrimary.currentPrice);
    const macdDiv = detectMACDDivergence(tfPrimary.candles.slice(-30), macdSeries.slice(-30));
    if (macdDiv.type === targetDivergence && macdDiv.strength >= 60) {
      score += 10; details.push(`${macdDiv.description} (strength ${macdDiv.strength})`);
    }
  }

  // RSI divergence (10%) - confirm timeframe
  if (tfConfirm.candles && tfConfirm.candles.length >= 20) {
    const rsiSeries = estimateRsiSeries(tfConfirm.candles, tfConfirm.rsi7);
    const rsiDiv = detectRSIDivergence(tfConfirm.candles.slice(-30), rsiSeries.slice(-30));
    if (rsiDiv.type === targetDivergence && rsiDiv.strength >= 60) {
      score += 10; details.push(`${rsiDiv.description} (strength ${rsiDiv.strength})`);
    }
  }

  const weakenedFrames = [primaryChange, confirmChange, filterChange].filter((c) => c.weakeningSeverity > 40).length;
  const hasDivergence = details.some((d) => d.includes("divergence"));
  const earlyWarning = weakenedFrames >= 2 || reversedFrames.length >= 2 || hasDivergence;

  let recommendation;
  if (score >= 70) recommendation = "close immediately - multiple timeframes confirm reversal";
  else if (score >= 50) recommendation = "recommend closing - reversal risk elevated";
  else if (earlyWarning && score >= 30) recommendation = "watch closely - trend weakening or divergence detected";
  else recommendation = "trend normal - continue holding";

  return { reversalScore: score, earlyWarning, timeframesReversed: reversedFrames, recommendation, details };
}

module.exports = {
  buildTimeframeIndicators,
  determineTrendStrength,
  determineMomentumState,
  determineVolatilityState,
  determineMarketState,
  calculateTrendConsistency,
  calculateTripleTimeframeConsistency,
  calculateTrendScore,
  detectTrendWeakening,
  getHistory,
  updateHistory,
  calculateReversalScore,
};

File: cloud-backtest/market_state_analyzer.py
"""
Faithful port of marketStateAnalyzer.js. Every threshold/formula copied
directly from the deployed JS source.

SCOPE NOTE (flagged, not hidden): the reversal score's MACD/RSI divergence
components (10%+10% of the JS version's weighting) are NOT ported here -
reconstructing them for a fast vectorized backtest adds real complexity for
a relatively small slice of the score. The primary/confirm/filter
trend-weakening components (80% of the weighting) ARE ported in full. This
means reversal-triggered exits in this backtest are somewhat less sensitive
than the live bot - if this matters for your conclusions, ask and it can be
added.
"""
import numpy as np
from indicators import ema, rsi, macd, macd_histogram_turn, atr_ratio, avg_volume, detect_volume_spike

OVERSOLD_EXTREME = 20
OVERSOLD_MILD = 30
OVERBOUGHT_EXTREME = 80
OVERBOUGHT_MILD = 70


def build_timeframe_indicators(candles):
    closes = candles["close"].values
    current_price = closes[-1]
    ema20 = ema(closes, 20)
    ema50 = ema(closes, 50)
    e20 = ema20[-1] if len(ema20) else 0
    e50 = ema50[-1] if len(ema50) else 0
    m = macd(closes) or {"macd": 0, "signal": 0, "histogram": 0}
    deviation = ((current_price - e20) / e20 * 100) if e20 != 0 else 0
    return {
        "current_price": current_price,
        "ema20": e20,
        "ema50": e50,
        "macd": m["macd"],
        "macd_signal": m["signal"],
        "macd_histogram": m["histogram"],
        "macd_turn": macd_histogram_turn(closes),
        "rsi7": rsi(closes, 7) if rsi(closes, 7) is not None else 50,
        "rsi14": rsi(closes, 14) if rsi(closes, 14) is not None else 50,
        "atr_ratio": atr_ratio(candles, 14),
        "deviation_from_ema20": deviation,
        "volume": candles["volume"].iloc[-1],
        "avg_volume": avg_volume(candles, 20),
        "candles": candles,
    }


def determine_trend_strength(tf):
    if tf["ema20"] > tf["ema50"] and tf["macd"] > 0:
        return "trending_up"
    if tf["ema20"] < tf["ema50"] and tf["macd"] < 0:
        return "trending_down"
    return "ranging"


def determine_momentum_state(tf):
    if tf["rsi7"] < OVERSOLD_EXTREME:
        return "oversold_extreme"
    if tf["rsi7"] < OVERSOLD_MILD:
        return "oversold_mild"
    if tf["rsi7"] > OVERBOUGHT_EXTREME:
        return "overbought_extreme"
    if tf["rsi7"] > OVERBOUGHT_MILD:
        return "overbought_mild"
    return "neutral"


def determine_market_state(trend_strength, momentum_state, tf_confirm):
    state = "no_clear_signal"
    confidence = 0.3

    if trend_strength == "trending_up" and momentum_state == "oversold_extreme":
        state, confidence = "uptrend_oversold", 0.9
    elif trend_strength == "trending_down" and momentum_state == "overbought_extreme":
        state, confidence = "downtrend_overbought", 0.9
    elif trend_strength == "trending_down" and momentum_state == "oversold_extreme":
        state, confidence = "downtrend_oversold", 0.6
    elif trend_strength == "trending_up" and momentum_state == "overbought_extreme":
        state, confidence = "uptrend_overbought", 0.6
    elif trend_strength == "trending_up" and momentum_state in ("oversold_mild", "neutral"):
        state, confidence = "uptrend_continuation", 0.7
    elif trend_strength == "trending_down" and momentum_state in ("overbought_mild", "neutral"):
        state, confidence = "downtrend_continuation", 0.7
    elif trend_strength == "trending_down" and momentum_state == "oversold_mild":
        state, confidence = "downtrend_oversold", 0.5
    elif trend_strength == "trending_up" and momentum_state == "overbought_mild":
        state, confidence = "uptrend_overbought", 0.5
    elif trend_strength == "ranging" and momentum_state == "oversold_extreme":
        state, confidence = "ranging_oversold", 0.8
    elif trend_strength == "ranging" and momentum_state == "overbought_extreme":
        state, confidence = "ranging_overbought", 0.8
    elif trend_strength == "ranging" and momentum_state == "neutral":
        state, confidence = "ranging_neutral", 0.5

    if tf_confirm["macd_turn"] == 1 and state in ("uptrend_oversold", "ranging_oversold"):
        confidence = min(confidence + 0.1, 1.0)
    if tf_confirm["macd_turn"] == -1 and state in ("downtrend_overbought", "ranging_overbought"):
        confidence = min(confidence + 0.1, 1.0)

    return {"state": state, "confidence": confidence}


def calculate_trend_consistency(ema20a, ema50a, ema20b, ema50b, macd_a, macd_b):
    score = 0
    trend_a = 1 if ema20a > ema50a else -1
    momentum_a = 1 if macd_a > 0 else -1
    trend_b = 1 if ema20b > ema50b else -1
    momentum_b = 1 if macd_b > 0 else -1
    if trend_a == trend_b:
        score += 0.4
    if momentum_a == momentum_b:
        score += 0.3
    if trend_a == momentum_a:
        score += 0.15
    if trend_b == momentum_b:
        score += 0.15
    return max(0, min(1, score))


def calculate_triple_timeframe_consistency(tf_primary, tf_confirm, tf_filter):
    pc = calculate_trend_consistency(
        tf_primary["ema20"], tf_primary["ema50"], tf_confirm["ema20"], tf_confirm["ema50"],
        tf_primary["macd"], tf_confirm["macd"])
    cf = calculate_trend_consistency(
        tf_confirm["ema20"], tf_confirm["ema50"], tf_filter["ema20"], tf_filter["ema50"],
        tf_confirm["macd"], tf_filter["macd"])
    return pc * 0.6 + cf * 0.4


def calculate_trend_score(tf):
    """-100..100. Matches source exactly: EMA gap (40%) + MACD/price
    normalized (30%) + price deviation from EMA20 (20%) + RSI trend (10%)."""
    score = 0
    ema_gap = (tf["ema20"] - tf["ema50"]) / tf["ema50"] if tf["ema50"] != 0 else 0
    score += max(-40, min(40, ema_gap * 1000))
    macd_normalized = tf["macd"] / tf["current_price"] if tf["current_price"] != 0 else 0
    score += max(-30, min(30, macd_normalized * 10000))
    score += max(-20, min(20, tf["deviation_from_ema20"] * 2))
    rsi_trend = (tf["rsi7"] - 50) / 5
    score += max(-10, min(10, rsi_trend))
    return round(score)


def detect_trend_weakening(current_score, score_history):
    """Matches source exactly: 20%-relative-drop weakening, +/-20 crossing
    = reversing."""
    previous_score = score_history[-1] if len(score_history) > 0 else current_score
    is_weakening = abs(current_score) < abs(previous_score) * 0.8
    weakening_severity = (
        round((1 - abs(current_score) / abs(previous_score)) * 100) if is_weakening and previous_score != 0 else 0
    )
    return {"current_score": current_score, "previous_score": previous_score, "is_weakening": is_weakening,
            "weakening_severity": weakening_severity}


def detect_adverse_drift(current_score, score_history, target_sign, net_threshold=10):
    """Faithful port of the JS fix (marketStateAnalyzer.js, added Aug 2 after
    the ETH/DOGE reversal-blind-spot incident). detect_trend_weakening above
    only fires on a single-step MAGNITUDE DROP - it's direction-blind, so a
    score climbing steadily toward the adverse side (14 -> 20 -> 22 -> 27,
    each step real, none of them a "drop") never registers. This looks at
    the last few cycles instead and asks a direction-based question: has the
    score moved toward the target (adverse) side on nearly every recent
    step, covering real cumulative distance - regardless of whether any
    single step was a big enough drop to trip the old check.

    net_threshold=10 was tuned against the real ETH short that triggered
    this fix (primary score 14/20/22/27, net +13 over 3 cycles) - NOT
    independently validated against 2 years of history yet. Treat this
    default as a hypothesis to backtest, not a settled number - see
    compare_adverse_drift.py.
    """
    sequence = list(score_history) + [current_score]
    if len(sequence) < 3:
        return {"is_drifting": False, "net_movement": 0, "adverse_steps": 0, "total_steps": 0}

    window = sequence[-4:]
    steps = [window[i] - window[i - 1] for i in range(1, len(window))]
    adverse_steps = sum(1 for s in steps if np.sign(s) == target_sign)
    net_movement = (window[-1] - window[0]) * target_sign

    is_drifting = (
        len(steps) >= 2
        and adverse_steps >= len(steps) - 1
        and net_movement >= net_threshold
        and np.sign(current_score) == target_sign
    )
    return {"is_drifting": bool(is_drifting), "net_movement": round(net_movement),
            "adverse_steps": adverse_steps, "total_steps": len(steps)}


def calculate_reversal_score(tf_primary, tf_confirm, tf_filter, position_direction, history,
                              use_adverse_drift=False, drift_net_threshold=10):
    """Weighted 40/25/15 across primary/confirm/filter timeframes (the
    trend-weakening portion of the source's reversal score - MACD/RSI
    divergence NOT included, see module docstring).

    use_adverse_drift: NEW, default False so every existing sweep/backtest
    call reproduces its old results unchanged unless explicitly opted in.
    When True, adds the drift branch (see detect_adverse_drift) to each
    timeframe's scoring ladder - this is the thing compare_adverse_drift.py
    is for: run the SAME data with this False vs True and compare.
    """
    score = 0
    details = []
    reversed_frames = []
    target_sign = -1 if position_direction == "long" else 1

    score_primary = calculate_trend_score(tf_primary)
    score_confirm = calculate_trend_score(tf_confirm)
    score_filter = calculate_trend_score(tf_filter)

    primary_change = detect_trend_weakening(score_primary, history.get("primary", []))
    primary_drift = detect_adverse_drift(score_primary, history.get("primary", []), target_sign, drift_net_threshold) if use_adverse_drift else None
    if np.sign(score_primary) == target_sign and abs(score_primary) > 30:
        score += 40
        details.append(f"primary strongly reversed ({score_primary})")
        reversed_frames.append("primary")
    elif primary_change["is_weakening"] and primary_change["weakening_severity"] > 40:
        score += 20
        details.append(f"primary weakening ({primary_change['weakening_severity']}%)")
    elif use_adverse_drift and primary_drift["is_drifting"]:
        score += 18
        details.append(f"primary drifting adverse (net {primary_drift['net_movement']})")
        reversed_frames.append("primary_drift")
    elif abs(score_primary) < 20:
        score += 12

    confirm_change = detect_trend_weakening(score_confirm, history.get("confirm", []))
    confirm_drift = detect_adverse_drift(score_confirm, history.get("confirm", []), target_sign, drift_net_threshold) if use_adverse_drift else None
    if np.sign(score_confirm) == target_sign and abs(score_confirm) > 30:
        score += 25
        details.append(f"confirm strongly reversed ({score_confirm})")
        reversed_frames.append("confirm")
    elif confirm_change["is_weakening"] and confirm_change["weakening_severity"] > 40:
        score += 12
    elif use_adverse_drift and confirm_drift["is_drifting"]:
        score += 11
        details.append(f"confirm drifting adverse (net {confirm_drift['net_movement']})")
        reversed_frames.append("confirm_drift")

    filter_change = detect_trend_weakening(score_filter, history.get("filter", []))
    filter_drift = detect_adverse_drift(score_filter, history.get("filter", []), target_sign, drift_net_threshold) if use_adverse_drift else None
    if np.sign(score_filter) == target_sign and abs(score_filter) > 30:
        score += 15
        details.append(f"filter reversed ({score_filter})")
        reversed_frames.append("filter")
    elif use_adverse_drift and filter_drift["is_drifting"]:
        score += 7
        details.append(f"filter drifting adverse (net {filter_drift['net_movement']})")
        reversed_frames.append("filter_drift")

    return {
        "reversal_score": score,
        "reversed_frames": reversed_frames,
        "details": details,
        "trend_scores": {"primary": score_primary, "confirm": score_confirm, "filter": score_filter},
    }
